### PR TITLE
Fix incompatibility with sails V1 

### PR DESCRIPTION
**Pull request by bzd2000**

### CODE REVIEW

**Comment by bzd2000** [comment]
I was not able to run the tests because of some issues with the test runner.
I did some basic unit testing on sails V1 (not on sails 0.12)

**Comment by edouardmenayde** [comment]
I think `sails-hook-authorization` should stay compatible to v12. We might make a big jump one day but v1.0 is not release yet and it seems a bit too early to make such decisions (I think you would agree @RWOverdijk).
I would advise you to create a second model compatible with v1.0 and load it if it's v1.0.
Same goes for your other fixes :-)
Besides that we have some guidelines about contributing you should [read](https://github.com/SpoonX/about/blob/master/CONTRIBUTING.md) :)

Ps:
That is the only way I found a the moment to test for sails version, it's pretty loose but it does the job.
```js
if (sails.registerAction) { // Test if V1
// V1
}
else {
 // < V1
}
```

File: api/controllers/AuthController.js [hunk 1]
```diff
@@ -17,10 +17,19 @@ module.exports = {
 
     params = params.asObject();
 
+    var finduser;
+
     if (authConfig.wetland) {
       findUser = req.getRepository(sails.models.user.Entity).findOne({[loginProperty]: params[loginProperty]}, {populate: populate});
     } else {
-      findUser = sails.models.user['findOneBy' + _.upperFirst(loginProperty)](params[loginProperty]);
+      sails.models.user.findOne({username: 'test'}).exec(function (err, user){
```
Comment (RWOverdijk) [inline api/controllers/AuthController.js]: `findUser = sails.models.user.findOne({[loginProperty]: params[loginProperty]});`

File: api/controllers/AuthController.js [hunk 2]
```diff
@@ -110,13 +119,16 @@ module.exports = {
...
     params = params.asObject();
+    var finduser;
 
     if (authConfig.wetland) {
       UserEntity = sails.models.user.Entity;
       manager    = req.getManager();
       findUser   = manager.getRepository(UserEntity).findOne({[loginProperty]: params[loginProperty]});
     } else {
-      findUser = sails.models.user['findOneBy' + _.upperFirst(loginProperty)](params[loginProperty]);
+      sails.models.user.findOne({username: 'test'}).exec(function (err, user){
```
Comment (RWOverdijk) [inline api/controllers/AuthController.js]: `findUser = sails.models.user.findOne({[loginProperty]: params[loginProperty]});`

Comment (RWOverdijk) [reply api/controllers/AuthController.js]: Undo the exec logic

File: api/controllers/AuthController.js (hunk 1 above)
Comment (RWOverdijk) [reply api/controllers/AuthController.js]: Undo the exec logic

File: api/controllers/AuthController.js [hunk 3]
```diff
@@ -17,10 +17,19 @@ module.exports = {
 
     params = params.asObject();
 
+    var finduser;
```
Comment (RWOverdijk) [inline api/controllers/AuthController.js]: Not needed

File: api/controllers/AuthController.js [hunk 4]
```diff
@@ -17,10 +17,19 @@ module.exports = {
...
+
     if (authConfig.wetland) {
       findUser = req.getRepository(sails.models.user.Entity).findOne({[loginProperty]: params[loginProperty]}, {populate: populate});
     } else {
-      findUser = sails.models.user['findOneBy' + _.upperFirst(loginProperty)](params[loginProperty]);
+      sails.models.user.findOne({username: 'test'}).exec(function (err, user){
+        finduser = user;
+      });
+
+          console.log ('findUser: ' + finduser);
```
Comment (RWOverdijk) [inline api/controllers/AuthController.js]: Remove (including the newlines surrounding it)

File: api/controllers/AuthController.js [hunk 5]
```diff
@@ -110,13 +119,16 @@ module.exports = {
     }
 
     params = params.asObject();
+    var finduser;
```
Comment (RWOverdijk) [inline api/controllers/AuthController.js]: Not needed

File: api/controllers/AuthController.js [hunk 6]
```diff
@@ -134,15 +146,18 @@ module.exports = {
...
       })
       .then(user => {
         let findUser;
 
         if (authConfig.wetland) {
           findUser = manager.getRepository(UserEntity).findOne(user.id, {populate: populate});
         } else {
-          findUser = sails.models.user.findOne(user.id);
+
+          sails.models.user.findOne(user.id).exec(function (err, user){
```
Comment (RWOverdijk) [inline api/controllers/AuthController.js]: Undo exec logic

File: api/controllers/AuthController.js [hunk 7]
```diff
@@ -205,7 +220,11 @@ module.exports = {
           return manager.getRepository(sails.models.user.Entity).findOne(decodedToken.activate);
         }
 
-        return sails.models.user.findOneId(decodedToken.activate);
+
+        sails.models.user.findOne(decodedToken.activate).exec(function (err, user){
```
Comment (RWOverdijk) [inline api/controllers/AuthController.js]: Undo exec logic

File: api/models/User.js [hunk 8]
```diff
@@ -1,15 +1,19 @@
 var bcrypt = require('bcrypt');
```
Comment (RWOverdijk) [inline api/models/User.js:1]: This file should be changed to something like:

`module.exports = parseFloat(sails.version) < 1 ? require('../lib/user.v1') : require('../lib/user.v2');`

File: api/models/User.js [hunk 9]
```diff
@@ -20,16 +24,20 @@ module.exports = {
     emailConfirmed: {
       type      : 'boolean',
       defaultsTo: false
-    },
+    }
+  },
 
-    toJSON: function() {
-      var values = this.toObject();
+  customToJSON: function() {
```
Comment (RWOverdijk) [inline api/models/User.js]: While moving (based on my other comment) look at indentation.

**Review by RWOverdijk — CHANGES_REQUESTED** [review]

**Review by bzd2000 — COMMENTED** [review]
Can you review it again?

I'm not sure about these lines
return sails.models.user.findOne(decodedToken.activate);

**Comment by bzd2000** [comment]
I'm not able to run the automated tests...
And I don't know how to import a local module into to a sails project.  So it is difficult to test it for me.
You know how to add a local module as depency in a sails project?

**Comment by RWOverdijk** [comment]
@bzd2000 The commit messages aren't properly formatted yet.

In any case, I'll review this when I have the time. I'd like to have extra eyes on this, so if @edouardmenayde and @VMBindraban could take a look as well that'd be great!

File: api/controllers/AuthController.js [hunk 10]
```diff
@@ -205,7 +207,9 @@ module.exports = {
           return manager.getRepository(sails.models.user.Entity).findOne(decodedToken.activate);
         }
 
-        return sails.models.user.findOneId(decodedToken.activate);
+
+        return sails.models.user.findOne(decodedToken.activate);
+
```
Comment (VMBindraban) [inline api/controllers/AuthController.js]: No need for the extra new-lines.

File: api/entity/User.js [hunk 11]
```diff
@@ -35,16 +35,15 @@ module.exports = class User extends Entity {
     }
 
     return bcrypt.hash(this.password, 10).then(hash => {
-      this.password = hash;
-    });
+      this.password = hash;    });
```
Comment (VMBindraban) [inline api/entity/User.js]: Format issue?

File: api/entity/User.js [hunk 12]
```diff
@@ -53,11 +52,11 @@ module.exports = class User extends Entity {
...
       bcrypt.getRounds(values.password);
     } catch(e) {
-      return bcrypt.hash(this.password, 10).then(hash => {
-        this.password = hash;
+      return bcrypt.hash(values.password, 10).then(hash => {
+        values.password = hash;
       });
     }
-  }
+  }
```
Comment (VMBindraban) [inline api/entity/User.js]: Remove space

File: api/models/User.js [hunk 13]
```diff
@@ -20,17 +23,14 @@ module.exports = {
...
-
-      delete values.password;
-
-      return values;
     }
   },
 
+  customToJSON: function() {
+  // Return a shallow copy of this record with the password.
+    return _.omit(this, ['password' ])
```
Comment (VMBindraban) [inline api/models/User.js:31]: Why not just call `toObject`?

File: test/bootstrap.wetland.test.js [hunk 14]
```diff
@@ -0,0 +1,52 @@
+    sails      = require('sails');
+var Barrels    = require('barrels');
+var path       = require('path');
```
Comment (VMBindraban) [inline test/bootstrap.wetland.test.js:3]: `const`?

**Review by VMBindraban — CHANGES_REQUESTED** [review]
Check for formatting, some weird things going on in the tests like alignment and newlines.

File: api/models/User.js (hunk 13 above)
Comment (bzd2000) [reply api/models/User.js:31]: It is like that in the sails doc

https://github.com/balderdashy/sails-docs/blob/1.0/upgrading/To1.0.md

The .toObject() instance method is no longer added to every record. When implementing customToJSON for a model, be sure to clone the record using _.omit(), _.pick() or _.clone().

File: test/bootstrap.wetland.test.js (hunk 14 above)
Comment (bzd2000) [reply test/bootstrap.wetland.test.js:3]: It was the only way I got it working.
With const it is also failing -> 

```
debug:  • https://sailsjs.com/support
debug: 
error: A hook (`policies`) failed to load!
  1) "before all" hook
error: A hook (`orm`) failed to load!

  0 passing (2s)
  1 failing

  1)  "before all" hook:
     ReferenceError: sails is not defined
      at Hook.<anonymous> (lib/loaderOverride.js:16:18)
      at Hook.wrapper [as loadPolicies] (node_modules/lodash/lodash.js:4968:19)
      at Hook.loadMiddleware (node_modules/sails/lib/hooks/policies/index.js:65:21)
      at Hook.wrapper [as loadMiddleware] (node_modules/@sailshq/lodash/lib/index.js:3250:19)
      at Hook.initialize (node_modules/sai
```

File: api/models/User.js (hunk 13 above)
Comment (VMBindraban) [reply api/models/User.js:31]: It is for sails < 1 however. I know bumping it to version 2.0 makes it a breaking change but do we still want to support sails 0.12? @RWOverdijk

Comment (bzd2000) [reply api/models/User.js:31]: But I thought that @RWOverdijk suggested to create a preview version for sails V1
To be checked with him

File: api/entity/User.js (hunk 12 above)
Comment (bzd2000) [reply api/entity/User.js]: fixed

File: api/entity/User.js (hunk 11 above)
Comment (bzd2000) [reply api/entity/User.js]: fixed

File: test/bootstrap.wetland.test.js (hunk 14 above)
Comment (bzd2000) [reply test/bootstrap.wetland.test.js:3]: But honestly ... I don't understand why

File: api/entity/User.js [hunk 15]
```diff
@@ -35,7 +35,7 @@ module.exports = class User extends Entity {
     }
 
     return bcrypt.hash(this.password, 10).then(hash => {
-      this.password = hash;
+      this.password = hash;
```
Comment (edouardmenayde) [inline api/entity/User.js]: Remove the trailing space :)

File: api/entity/User.js [hunk 16]
```diff
@@ -53,11 +53,11 @@ module.exports = class User extends Entity {
...
       bcrypt.getRounds(values.password);
     } catch(e) {
-      return bcrypt.hash(this.password, 10).then(hash => {
-        this.password = hash;
+      return bcrypt.hash(values.password, 10).then(hash => {
+        values.password = hash;
       });
     }
-  }
+  }
```
Comment (edouardmenayde) [inline api/entity/User.js]: Remove the trailing space

**Review by edouardmenayde — CHANGES_REQUESTED** [review]
I think the PR is on a good way, that's great. I didn't comment the test yet at the look of it you are still changing them a lot.

File: api/entity/User.js (hunk 15 above)
Comment (bzd2000) [reply api/entity/User.js]: done

File: api/entity/User.js (hunk 16 above)
Comment (bzd2000) [reply api/entity/User.js]: done

**Comment by bzd2000** [comment]
Did not intent to change the tests ... is there something still incorrect or missing?

File: test/bootstrap.wetland.test.js [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+  const migrator = sails.wetland.getMigrator();
+  const cleaner = sails.wetland.getCleaner();
+
+  cleaner.clean() // Will clean the database, NO MAGICAL GOING BACK
+      .then(() => migrator.devMigrations(false)) // Will actually do the migrations : needed here because the clean method wipes the database entirely
+      .then(() => done(err, server)) ;// Will seed accordingly to the configuration you gave wetland
+
+
```
Comment (edouardmenayde) [inline test/bootstrap.wetland.test.js]: Too much break lines

File: test/integration/controllers/AuthController.test.js [hunk 18]
```diff
@@ -118,4 +130,75 @@ describe('AuthController', () => {
...
+      // udpate
+      request(sails.hooks.http.app)
+        .patch('/user/' + userIdToUpdate)
+        .send(updateCredentials)
+        .expect(200, (error, response) => {
+
+          var body = response.body;
+          var body                = response.body;
+
+
```
Comment (edouardmenayde) [inline test/integration/controllers/AuthController.test.js]: Remove the line

File: test/integration/controllers/AuthController.test.js [hunk 19]
```diff
@@ -118,4 +130,75 @@ describe('AuthController', () => {
...
+        .post('/auth/login')
+        .send(updateCredentials)
+        .expect(200, (error, response) => {
+
+          var body = response.body;
+
+          assert.isDefined(body.access_token);
+          assert.isDefined(body.refresh_token);
+
+
```
Comment (edouardmenayde) [inline test/integration/controllers/AuthController.test.js]: Too much break lines

**Review by edouardmenayde — CHANGES_REQUESTED** [review]

**Comment by edouardmenayde** [comment]
@bzd2000 Not a lot in fact, I thought I saw commented code but nope.

File: test/integration/controllers/AuthController.test.js (hunk 19 above)
Comment (bzd2000) [reply test/integration/controllers/AuthController.test.js]: done

File: test/integration/controllers/AuthController.test.js (hunk 18 above)
Comment (bzd2000) [reply test/integration/controllers/AuthController.test.js]: done

File: test/bootstrap.wetland.test.js (hunk 17 above)
Comment (bzd2000) [reply test/bootstrap.wetland.test.js]: done

**Comment by RWOverdijk** [comment]
@bzd2000 It's not a huge deal (I can fix it afterwards) but your commit messages have a lot of fixes in it, and use a space.

Example: `fix (test): remove empty lines` should be `style(test): remove excessive newlines`

Other than that, I'll take some time to rereview this today. @edouardmenayde and @VMBindraban you're happy with the rest? Ergo, if I finish my review and everything's fine, you're ok with me merging?

**Comment by edouardmenayde** [comment]
I'm ok with the merging, I just have one little suggestion : it would be cool to add a travis config so that benefit from those tests but I guess that would be better suited in a new issue.

**Comment by RWOverdijk** [comment]
Hey, I was still reviewing this 😇

**Pull request closed without merging**